xformers/components/attention/core.py

```diff
@@ -231,30 +234,23 @@ def scaled_dot_product_attention(
         and q.shape[-2] == k.shape[-2]
         and q.shape[-2] < att_mask.shape[1]
     ):
-        seq = q.shape[-2]
-        if att_mask.ndim == 2:
-
-            if not att_mask.is_sparse:
-                att_mask = att_mask[:seq, :seq]
-            else:
-                logging.warning(
-                    "Mismatching attention mask and sequence length. On the fly correction but this will be slow"
-                )
-                # Loosing sparsity on purpose,
-                # expectation is that moving back and forth dense/sparse will negate the speedup
-                att_mask = att_mask.to_dense().squeeze(0)[:seq, :seq]
+        if isinstance(att_mask, AttentionMask):
```

Comment:
this is cleaner than before I believe, but the dense/sparse case is exposed

Reply:
I still believe we should not be cropping the mask in our core functions. But ok to keep this as is for now

Reply:
it's also a matter of deduplicating, else it's something which has to be done in all the corresponding attentions. No strong opinion though, I'm ok to move that up in SDP and others

Reply:
fixed that @fmassa :)